### DISCUSSION TITLE
CVRP with loading and unloading different order types

### DISCUSSION THREAD

**Discussion by FernandoDeAlmeidaAc** [google/or-tools#3749] [Routing (and legacy CP) questions]
Hi everyone.

I am having a problem while solving a CVRP with loading and unloading different order types that can be described in the example bellow:

Let’s say we have three nodes (Node1, Node2 and Node3). Node1 needs to return 2 blue boxes to the depot. Both Node2 and Node3 need to receive 1 yellow box from the depot. There are two identical vehicles available. Each vehicle can carry a maximum of 3 boxes (blue and/or yellow). So, we have:

Demands[blue] = [2, 0, 0]
Demands[yellow] = [0, -1, -1]

I cannot just create a dimension to deal with demands globally, because I would have 

Demands[global] = [2, -1, -1] 

and the route 

{Depot load(0) -> Node1 load(2) -> Node2 load(1) -> Node3 load(0) -> Depot load(0) } 

would be feasible. However, as the two blue boxes from Node1 cannot be used to service Node2 and Node3, the actual load record for this route must be

{Depot load(2) -> Node1 load(4) -> Node2 load(3) -> Node3 load(2) -> Depot load(2) }.

and since this route would violate the capacity of the vehicle, it actually is infeasible. I would like to also create a dimension for each individual demand and somehow make

dimensionGlobal = dimensionBlue + dimensionYellow. 

I don’t know if this is possible. If it’s possible, how can I do it? If it’s not, what else can I do?

I'm working with the c++ library version. Thanks.

**Comment by jmarca** [comment]
If the boxes are identical, the easiest solution is to create n+1
dimensions, where n is the number of colors, and +1 is a generic "box"
dimension.

A box always fills the box dimension.

A colored box only fills its own color.

So, in your example, the Demands are tuples at each location.  I would
write it as a hash of hashes, but I'm sure you can figure out the equivalent c++:

```python
Demands = {'Node1': {'blue': 2, 'yellow': 0, 'box': 2},
           'Node2': {'blue': 0, 'yellow': -1, 'box': -1},
           'Node3': {'blue': 0, 'yellow': -1, 'box': -1},
           }
```

vehicle capacities are defined for each dimension, as usual, and the
solver should correctly manage the limits---if the max load is 10 of
any box, the "box" dimension will prevent loading 10 yellow *and* 10
blue at the same time, as that would require 20 "box".

The hard part is when the boxes are not identical, or rather, when
they combine in non-linear ways.  If carrying any number of red boxes
means you can only load one yellow box, that sort of thing.

Good luck,

James

**Reply by FernandoDeAlmeidaAc** [reply]
But the solver needs to "know" that the generic "box" dimension is related to the color dimensions. Otherwise, it will simply treat it as another type of box, that is, it will mistakenly use the 2 generic boxes from Node1 to meet the generic demands of Node2 and Node3.

**Comment by jmarca** [comment]
The solver knows nothing. You tell it with the callback in the dimension definitions. 

Yellow dimension looks up the demand for yellow. 

Box dimension looks up the demand for box. Try it and it will become obvious. 

James

**Reply by FernandoDeAlmeidaAc** [reply]
I tried and the result was exactly as I said.  The 2 generic boxes from Node1 were used to meet the generic demands of Node2 and Node3. For this to work, the generic box dimension needs a callback that somehow gets the sum of the color dimensions.

**Reply by FernandoDeAlmeidaAc** [reply]
```cpp
#include <cstdint>
#include <sstream>
#include <string>
#include <utility>
#include <vector>
#include <string>

#include "ortools/constraint_solver/routing.h"
#include "ortools/constraint_solver/routing_enums.pb.h"
#include "ortools/constraint_solver/routing_index_manager.h"
#include "ortools/constraint_solver/routing_parameters.h"

namespace operations_research {
    struct DataModel {
        const std::vector<std::vector<int64_t>> distance_matrix{
            {0, 696, 548, 776},
            {696, 0, 308, 992},
            {548, 308, 0, 684},
            {776, 992, 684, 0},
        };
        const std::vector<std::vector<int64_t>> demands{
            //0   1   2  3
            { 0,  2,  0,  0}, //  blue box
            { 0,  0, -1, -1}, //  yellow box
            { 0,  2, -1, -1}, //  "generic" box
        };
        const std::vector<int64_t> vehicle_capacities{ 3, 3 };
        const int num_vehicles = 2;
        const RoutingIndexManager::NodeIndex depot{ 0 };
    };

    //! @brief Print the solution.
    //! @param[in] data Data of the problem.
    //! @param[in] manager Index manager used.
    //! @param[in] routing Routing solver used.
    //! @param[in] solution Solution found by the solver.
    void PrintResults(const DataModel& data, const RoutingIndexManager& manager,
        const RoutingModel& routing, const Assignment& solution) {
        int64_t total_distance{ 0 };
        int64_t total_load{ 0 };
        const size_t INDEX{2}; // SET THE LOAD TYPE {0, 1, or 2} TO BE PRINTED HERE!!!
        RoutingDimension* dimension = routing.GetMutableDimension("Capacity" + std::to_string(INDEX));
        for (int vehicle_id = 0; vehicle_id < data.num_vehicles; ++vehicle_id) {
            int64_t index = routing.Start(vehicle_id);
            LOG(INFO) << "Route for Vehicle " << vehicle_id << ":";
            int64_t route_distance{ 0 };
            int64_t route_load{ 0 };
            std::stringstream route;
            while (routing.IsEnd(index) == false)
            {
                int64_t node_index = manager.IndexToNode(index).value();
                route_load += data.demands[INDEX][node_index];
                int64_t previous_index = index;
                index = solution.Value(routing.NextVar(index));
                route << " -> " << node_index << " Load("
                    << solution.Value(dimension->CumulVar(index)) << ")";
                route_distance += routing.GetArcCostForVehicle(previous_index, index,
                    int64_t{ vehicle_id });
            };

            LOG(INFO) << route.str() << " -> " << manager.IndexToNode(index).value();
            LOG(INFO) << "Distance of the route: " << route_distance << "m";
            LOG(INFO) << "Load of the route: " << route_load;
            total_distance += route_distance;
            total_load += route_load;
        }
        LOG(INFO) << "Total distance of all routes: " << total_distance << "m";
        LOG(INFO) << "Total load of all routes: " << total_load;
        LOG(INFO) << "";
        LOG(INFO) << "Advanced usage:";
        LOG(INFO) << "Problem solved in " << routing.solver()->wall_time() << "ms";
    }

    void PrintSolution(const DataModel& data, const RoutingIndexManager& manager,
        const RoutingModel& routing, const Assignment& solution)
    {
        std::tuple<double, double> result = std::make_tuple(0.0, 0.0);

        switch (routing.status())
        {
        case RoutingModel::ROUTING_SUCCESS:
            LOG(INFO) << "Problem solved successfully after calling RoutingModel::Solve().";
            PrintResults(data, manager, routing, solution);
            break;
        case RoutingModel::ROUTING_PARTIAL_SUCCESS_LOCAL_OPTIMUM_NOT_REACHED:
            LOG(INFO) << "Problem solved successfully after calling RoutingModel::Solve(), except "
                << "that a local optimum has not been reached. Leaving more time would allow "
                << "improving the solution.";
            PrintResults(data, manager, routing, solution);
            break;
        case RoutingModel::ROUTING_INFEASIBLE:
            LOG(INFO) << "The problem was proven infeasible.";
            break;
        case RoutingModel::ROUTING_INVALID:
            LOG(INFO) << "Model, model parameters or flags are not valid. ";
            break;
        case RoutingModel::ROUTING_FAIL:
            LOG(INFO) << "No solution found to the problem after calling RoutingModel::Solve().";
            break;
        case RoutingModel::ROUTING_FAIL_TIMEOUT:
            LOG(INFO) << "Time limit reached before finding a solution with RoutingModel::Solve().";
            break;
        case RoutingModel::ROUTING_NOT_SOLVED:
            LOG(INFO) << "Problem not solved yet (before calling RoutingModel::Solve()).";
            break;
        default:
            LOG(INFO) << "Undefined Status: " << routing.status();
            break;
        }
    }

    void VrpCapacity() {
        // Instantiate the data problem.
        DataModel data;

        // Create Routing Index Manager
        RoutingIndexManager manager(data.distance_matrix.size(), data.num_vehicles,
            data.depot);

        // Create Routing Model.
        RoutingModel routing(manager);

        // Create and register a transit callback.
        const int transit_callback_index = routing.RegisterTransitCallback(
            [&data, &manager](int64_t from_index, int64_t to_index) -> int64_t {
                // Convert from routing variable Index to distance matrix NodeIndex.
                int from_node = manager.IndexToNode(from_index).value();
        int to_node = manager.IndexToNode(to_index).value();
        return data.distance_matrix[from_node][to_node];
            });

        // Define cost of each arc.
        routing.SetArcCostEvaluatorOfAllVehicles(transit_callback_index);

        // Add Capacity constraint.
        for (size_t i = 0; i < data.demands.size(); i++)
        {
            const int demand_callback_index = routing.RegisterUnaryTransitCallback(
                [&data, &manager, i](int64_t from_index) -> int64_t {
                    // Convert from routing variable Index to demand NodeIndex.
                    int from_node = manager.IndexToNode(from_index).value();
            return data.demands[i][from_node];
                });
            routing.AddDimensionWithVehicleCapacity(
                demand_callback_index,    // transit callback index
                int64_t{ 0 },               // capacity slack
                data.vehicle_capacities,  // vehicle maximum capacities
                false,                     // start cumul to zero
                "Capacity" + std::to_string(i));
        }

        // Setting first solution heuristic.
        RoutingSearchParameters search_parameters = DefaultRoutingSearchParameters();
        search_parameters.set_first_solution_strategy(
            FirstSolutionStrategy::PATH_CHEAPEST_ARC);
        search_parameters.set_local_search_metaheuristic(
            LocalSearchMetaheuristic::GUIDED_LOCAL_SEARCH);
        search_parameters.mutable_time_limit()->set_seconds(1);

        // Solve the problem.
        const Assignment* solution = routing.SolveWithParameters(search_parameters);

        // Print solution on console.
        PrintSolution(data, manager, routing, *solution);
    }
}

int main(int /*argc*/, char* /*argv*/[]) {
    operations_research::VrpCapacity();
    return EXIT_SUCCESS;
}
```

**Comment by jmarca** [comment]
Something is not right then.  I should not be possible to provide 2 yellow boxes if you have 0 available in that dimension.  

Staring at the code, you have

```
        const std::vector<std::vector<int64_t>> demands{
            //0   1   2  3
            { 0,  2,  0,  0}, //  blue box
            { 0,  0, -1, -1}, //  yellow box
            { 0,  2, -1, -1}, //  "generic" box
        };
```

That seems fine, although I'm not sure about the depot having 0 values.

Then in the callback you do:

```
return data.demands[i][from_node];
```
which seems right.  

When you visit node 1,  the demand[0][1] return 2 (blue box) , the demand[1][1] returns 0 (yellow box), and the demand[2][1] returns 2 (box).  Then if it next visits node 2, it should fail, because the loads cannot go below 0.  The demand[0][2] is 0, so that is fine, and leaves you with blue=2.  The demand[1][2] is -1, and that should fail, because ... well no.  Perhaps it won't fail.  I recall that if you push a dimension below 0, it will just silently truncate to 0.

Oh, you also have "start cumul to zero" set to false.  Perhaps that is what is going on?  It can load up magically 2 yellow boxes, then 
deliver them.  

Can you print out the actual load of each dimension in your results at each node?

**Reply by FernandoDeAlmeidaAc** [reply]
Blue box load
`const size_t INDEX{0}; // SET THE LOAD TYPE {0, 1, or 2} TO BE PRINTED HERE!!!`

<img width="950" alt="image" src="https://user-images.githubusercontent.com/128821293/232891191-7b586aa5-ffab-4c8e-a970-32e80c3c11e7.png">

Yellow box load
`const size_t INDEX{1}; // SET THE LOAD TYPE {0, 1, or 2} TO BE PRINTED HERE!!!`

<img width="944" alt="image" src="https://user-images.githubusercontent.com/128821293/232891965-e95b3a64-60b9-478f-b214-2674f01b0c55.png">

"Generic" box load
`const size_t INDEX{2}; // SET THE LOAD TYPE {0, 1, or 2} TO BE PRINTED HERE!!!`

<img width="956" alt="image" src="https://user-images.githubusercontent.com/128821293/232892418-42cb09b9-aa15-4dde-aeb7-d47ad282eb67.png">

I set "start cumul to zero" to **false** to allow the vehicle to pick up boxes from the depot if needed. This happened in the case of the yellow loading to prevent the dimension from reaching a negative number.

**Comment by jmarca** [comment]
Wait, I'm confused.  Why aren't they all solved at the same time?  Why are there three different runs?  And the bug is that depot node allowing non-zero start volume.  That breaks the assumption that generic box  is the sum of all boxes.  So you load 2 yellow boxes, but do not have the equivalent in the box dimension.

Get rid of that, and be more specific about the pickups and deliveries.

**Reply by FernandoDeAlmeidaAc** [reply]
They were resolved at the same time, I just wasn't printing them together. I've now modified the PrintResults function to do just that.

```cpp
    void PrintResults(const DataModel& data, const RoutingIndexManager& manager,
        const RoutingModel& routing, const Assignment& solution) {
        int64_t total_distance{ 0 };
        int64_t total_load{ 0 };
        RoutingDimension* blue = routing.GetMutableDimension("Capacity0");
        RoutingDimension* yellow = routing.GetMutableDimension("Capacity1");
        RoutingDimension* generic = routing.GetMutableDimension("Capacity2");
        for (int vehicle_id = 0; vehicle_id < data.num_vehicles; ++vehicle_id) {
            int64_t index = routing.Start(vehicle_id);
            LOG(INFO) << "Route for Vehicle " << vehicle_id << ":";
            int64_t route_distance{ 0 };
            int64_t route_load{ 0 };
            std::stringstream route;
            while (routing.IsEnd(index) == false)
            {
                int64_t node_index = manager.IndexToNode(index).value();
                route_load += data.demands[2][node_index];
                int64_t previous_index = index;
                index = solution.Value(routing.NextVar(index));
                route << " -> " << node_index
                    << " B(" << solution.Value(blue->CumulVar(index)) << ")"
                    << "Y(" << solution.Value(yellow->CumulVar(index)) << ")"
                    << "G(" << solution.Value(generic->CumulVar(index)) << ")";
                route_distance += routing.GetArcCostForVehicle(previous_index, index,
                    int64_t{ vehicle_id });
            };

            LOG(INFO) << route.str() << " -> " << manager.IndexToNode(index).value();
            LOG(INFO) << "Distance of the route: " << route_distance << "m";
            LOG(INFO) << "Load of the route: " << route_load;
            total_distance += route_distance;
            total_load += route_load;
        }
        LOG(INFO) << "Total distance of all routes: " << total_distance << "m";
        LOG(INFO) << "Total load of all routes: " << total_load;
        LOG(INFO) << "";
        LOG(INFO) << "Advanced usage:";
        LOG(INFO) << "Problem solved in " << routing.solver()->wall_time() << "ms";
    }
```
<img width="858" alt="image" src="https://user-images.githubusercontent.com/128821293/233069995-70f08e22-41f5-4dfc-8002-f940dee3c0bf.png">

In theory, if I just "get rid of it", i.e. if I set "start cumul to zero" to **true**, I would have
-> 0 B(0)Y(0)G(0) -> 1 B(2)Y(0)G(2) -> 2 B(2)Y(**-1**)G(1) -> 3 B(2)Y(**-2**)G(0) -> 0.
But in practice I get an unfeasible return because no dimension is allowed to reach a negative number.

<img width="853" alt="image" src="https://user-images.githubusercontent.com/128821293/233072494-0a4870a0-d7a6-4994-8b3a-4400e16ed4ae.png">

I set it to "false" to make the depot an infinite source of boxes that can be used if needed. Because, as said before, Node2 and Node3 need to receive 1 yellow box **from the depot**.

In theory, if somehow, I could get the relationship between the dimensions (and not just between the input data) to be _generic = blue + yellow_ in this condition (start_cumul_to_zero = false) I would have
-> 0 B(0)Y(2)G(2) -> 1 B(2)Y(2)G(**4**) -> 2 B(2)Y(1)G(**3**) -> 3 B(2)Y(0)G(2) -> 0.
But this would violate the vehicle's capability, and therefore the solver would be forced to use the second vehicle to produce a solution. This is what I want.

**Comment by jmarca** [comment]
Following up.  In my humble opinion, letting the dimension value float at start is a hack that should only be used for simple problems.  When you start to deal with more realistic problems (like differentiated items to deliver, etc), then you need to use a proper pickup and deliver formulation, where every item is picked up and then delivered.  This is more complicated to set up, but is necessary.  

Here that would be one pickup each for each of the yellow box deliveries at node 2 and 3, and one delivery for the two blue box pickups at node 1.

The pickups and deliveries occur at the depot, so you'd make dummy nodes there for each.

**Reply by FernandoDeAlmeidaAc** [reply]
I don't like the idea of creating dummy nodes for each one. And I would still have to create dummy loads to prevent the dimensions from reaching negative numbers. It doesn't seem to be practical for large realistic problems.

**Comment by FernandoDeAlmeidaAc** [comment]
I found a solution.  I added the constraint _blue + yellow = generic_ to routing.solver() for every route start node.

```cpp
        Solver* solver = routing.solver();
        RoutingDimension* dimension_blue = routing.GetMutableDimension("Capacity0");
        RoutingDimension* dimension_yellow = routing.GetMutableDimension("Capacity1");
        RoutingDimension* dimension_global = routing.GetMutableDimension("Capacity2");

        for (size_t i = 0; i < data.num_vehicles; i++)
        {
            solver->AddConstraint(
                solver->MakeEquality( // 
                    solver->MakeSum({ dimension_blue->CumulVar(routing.Start(i)), dimension_yellow->CumulVar(routing.Start(i)) }),
                    dimension_global->CumulVar(routing.Start(i))
                )
            );
        }
```

The solver found that changing the route sequence is cheaper than adding the second vehicle, but the load issue was resolved anyway.

<img width="854" alt="image" src="https://user-images.githubusercontent.com/128821293/233136516-68c8e486-24b9-4c94-925d-f87581ed8707.png">

**Comment by jmarca** [comment]
Nice. I was just going to suggest that, but you did the link better than I would have. 

Still, I really hate relying on magic behavior, and I’ve found linking dimensions inside the solver to be a performance hit. 

I wonder about the performance difference between this approach and my explicit PDPTW approach with a larger problem, say 500 nodes, 50 vehicles, 10 item types. 

James
